### ISSUE TITLE
Labels for inputs on pretixpresale/event/index.html

### ISSUE THREAD

**Issue by RaphaelKimmig** [pretix/pretix#1128]
We've had feedback from multiple users that they were confused by the unlabelled input fields for item counts on the ticket page (see e.g. https://pretix.eu/demo/democon/). The issue may be more pronounced with multiple tickets with longer descriptions.

The fact that a user has to enter a number into that field to order a ticket isn't evident and the purpose of the field isn't visible anywhere on the page. Using the current design one could for example misuse the place holder as a label or maybe add an input group add-on to the count input as well.

![Examples with the current design](https://user-images.githubusercontent.com/226708/50910019-411cc880-142d-11e9-8628-1493cb078a43.png)

**Comment by HDValentin** [comment]
I had last year some contact with some confused people about the missing Label. They all did not understand to rise the amount of tickets to get the "put to basket"-button enabled.

I think it could be an accessibility issue for blind users too.

**Comment by rixx** [comment]
I'm not sure a label would help all that much. I've run into this issue on the support side repeatedly, too. When I thought about the problem, I always thought that direct action buttons per input would be nice.
![deepinscreenshot_select-area_20190223213059](https://user-images.githubusercontent.com/2657547/53291370-5e75de00-37b2-11e9-8dab-c6eb2aad99cd.png)
![deepinscreenshot_select-area_20190223213021](https://user-images.githubusercontent.com/2657547/53291371-5e75de00-37b2-11e9-8952-f2e820158c8b.png)

**Comment by raphaelm** [comment]
Hmm, that looks pretty, but I see two downsides with individual actions:

* It gets crowded quickly. For English, "to Cart" is still fine, but a German "in den Warenkorb" already requires more space than we really have for both the input field and the button.

* It changes the semantics quite heavily, as it would no longer be possible to add multiple things to the cart at once. Even if we would keep the "global" button as an option, it would then be very inconvenient that everything is filled with a "1" by default.

I think replacing the placeholder "0" with "Number"/"Anzahl" already helps a little, but maybe we do need a more radical change like this one. Happy for more ideas :)

**Comment by HDValentin** [comment]
It is necessary to have a label for every input field!
Because of accessibility reasons. Every tutorial starts with labels:

https://www.barrierefreies-webdesign.de/knowhow/formulare/label.html
https://webaim.org/techniques/forms/
https://www.w3.org/WAI/tutorials/forms/
https://medium.com/net-magazine/create-simple-accessible-web-forms-535cd253dee2
https://blog.prototypr.io/designing-accessible-forms-82f2ea11697f
https://uxcellence.com/2018/accessible-forms

**Comment by raphaelm** [comment]
Hi Valentin, do not fear, our fields do have labels. Although it has been a while since I read a web tutorial, I've been doing this for a while now ;) This issue is not at all about what HTML calls a label, i.e. ``<label>`` tags, but about *visually* labeling the user interface.

Accessibility is explicitly **not** the concern of this issue. The field **does** already have a ``<label>``, which is the name of the product (e.g. "Standard Ticket" in the screenshot above). Additionally, the input fields have a ``title`` attribute with a content like "Amount of Standard Ticket to order". I believe this provides all required information to screenreader users to understand what's being asked for. To exaggerate, the point of this issue might even be that it's *harder* for visual users than for screen reader users to figure this out.

What we need to solve is how to provide a *visual* clue that an amount is requested here, without cluttering the interface.

**Comment by RaphaelKimmig** [comment]
I agree that adding buttons everywhere might make things confusing in other ways. I also feel like there might be a nicer solution by rethinking the design, however I didn't have any really good ideas, so far.

One thing that might also help would be adding a popover to the button when it is disabled.
Combining this with labels as placeholders might help improve the solution until a more comprehensive redesign is done.
![popover_disabled](https://user-images.githubusercontent.com/226708/53351865-15dd3280-3922-11e9-9e37-3ac121a4e9a0.png)

**Comment by raphaelm** [comment]
> One thing that might also help would be adding a popover to the button when it is disabled.
> Combining this with labels as placeholders might help improve the solution until a more comprehensive redesign is done.

That's something we can do anyway, yes, good idea!

**Comment by HDValentin** [comment]
Sorry for the misunderstanding and my unqualified answer. I think I should read some of the tutorials.

**Comment by raphaelm** [comment]
Don't worry about it, these things happen! :)

**Comment by raphaelm** [comment]
For now, I added the popover as @RaphaelKimmig suggested. feel free to re-open this if you still run into users not finding their way!

**Comment by HDValentin** [comment]
The Popover is better than nothing.
But it is no solution for a mobile device. There is no popup.

**Comment by raphaelm** [comment]
We haven't deployed the popover on pretix.eu yet (will happen during the day today), but that's a good point, I'll test if we can make it work on mobile.

**Comment by HDValentin** [comment]
This issue is confusing people.
I myself had just the problem.
Labels would here be great.
Like here mentioned or on top of the list https://github.com/pretix/pretix/issues/1128#issue-397429218

Another help would be the possibility to switch the places of the price field and the Amount field.
In nearly every bill and shop the amount is first and the price is on second place.
**Is it possible to implement an option to change the fields in the view here?
This would directly improve the usability.**

![pretix Switch-Fields](https://user-images.githubusercontent.com/1276955/90044526-0ac0c280-dcce-11ea-8a25-4d0990b0ca07.png)

**Comment by raphaelm** [comment]
Sorry, no, we will not switch the two fields since prices being an input field is a rarity in itself (in most cases prices are fixed, and it would look really weird the other way round), and making the order of the fields configurable just adds too much complexity.

**Comment by HDValentin** [comment]
@raphaelm I thought it would be just a change of view thing, sorry.
What about the missing Labels/Heading for the fields?
Because: people **are** doing it wrong. I have seen and experienced it.
Then they get an error, don't read it and leave the ticketshop.

**Comment by raphaelm** [comment]
I think we can close this since (a) swapping the fields is not going to happen at least not outside the scope of a full page redesign, (b) the issue with the add-to-cart button is now tracked in #2210 and (c) proper labels for screenreader users are happening in #2081

**Comment by wiffbi** [comment]
I think the labels being invisible is still a valid concern regarding usability in general. For screenreader users #2081 adds those with an improved wording (e.g. adding „quantity“), but for „regular/visual“ users visible labels could also be helpful. I also understand the current pattern of entering the quantity for all tickets I want and then adding all of them at once to the cart by hitting the button „add to cart“ at the bottom is not the most common pattern, which makes it even more important for the user to understand, what to do with the inputs. I myself stumbled upon this when I started with pretix.

For events with lots of products this also means that the add-to-cart button is not visible and it might not immediately be clear to the user what to do next after entering the quantity for a product/ticket. This could be solved with lots of add-to-basket buttons (as in the screenshot posted by @rixx either using an icon "+basket" or similar to minimise space usage when all of them are shown or having the quantity set to 0 and the add-to-cart-button only appear when a quantity>0 is entered) or having the global add-to-cart button always in view by using position:sticky in some way to scroll along when the form is visible.

Another totally different idea came to my mind regarding quantity-inputs: instead of free range number inputs maybe select-elements (with e.g. a range of 1-10) would work as well? We could use an initial `<option value="" selected>Quantity</option>` which makes it easier to understand for sighted users. Blind users still get the hidden aria-labels.
Using `<select>` would also „fix“ an a11y issue regarding those rocker-buttons present in input type=number fields. During A11y-testing these where identified as a barrier (I won’t go into details here, but after a long discussion I understand and respect the issues they brought forward regarding those rocker-buttons).

**Comment by raphaelm** [comment]
That <select> approach does no longer works very well for shops where people buy tickets for groups of 100, which do exist. Otherwise it's not that uncommon with our competitors.

**Comment by raphaelm** [comment]
Another option to the below-fold button could even be having the button sticky to the bottom of the screen, but that would require a more intense design change probably. Could work out nice, though.

**Comment by wiffbi** [comment]
Regarding the sticky add-to-cart button, I’d stick it to the bottom of the page, but confine in the product-listing container, so it does not move to far away from the product list when scrolling down and reaching the footer-section with voucher entry, etc.

I agree, <select> does not work for more than 10 options. The question is, whether this rare use case of 100 tickets and more should set the default for all others or if we can find some kind of compromise. I currently do not have a clever idea though.

Regarding labels, I added screenshots to #2211 which might be usefull here as well. My favorite so far would be this:
![pretix-products-table](https://user-images.githubusercontent.com/276509/135612396-ac880ebb-db44-4fe7-9d18-d22d74628818.png)

**Comment by wiffbi** [comment]
Testing the sticky-button approach, I realised the aforementioned table-like „labels“ does not translate to mobile at all. So maybe instead of these labels, we could explore the `<select>` approach with `<option value="0">Quantity</option>` by introducing an `<option>more …</option>` at the end of the list, which basically replaces the select-element with a number-input. That way we get a visual „label“ for the user to understand what to do with the input and still make it possible to order 100+ tickets easily.

Regarding the sticky add-to-cart button: adter testing it, I do not think it works so well or even at all. To me it seems more confusing as a user. To see for yourself, I added a WIP PR #2237 .
